sway-core/src/semantic_analysis/program.rs

```diff
@@ -1,23 +1,228 @@
 };
 
+fn call_encode(_engines: &Engines, arg: Expression) -> Expression {
+    Expression {
+        kind: ExpressionKind::FunctionApplication(Box::new(FunctionApplicationExpression {
+            call_path_binding: TypeBinding {
+                inner: CallPath {
+                    prefixes: vec![],
+                    suffix: Ident::new_no_span("encode".into()),
+                    is_absolute: false,
+                },
+                type_arguments: TypeArgs::Regular(vec![]),
+                span: Span::dummy(),
+            },
+            arguments: vec![arg],
+        })),
+        span: Span::dummy(),
+    }
+}
+
+fn call_decode_first_param(engines: &Engines) -> Expression {
+    let string_slice_type_id = engines.te().insert(engines, TypeInfo::StringSlice, None);
+    Expression {
+        kind: ExpressionKind::FunctionApplication(Box::new(FunctionApplicationExpression {
+            call_path_binding: TypeBinding {
+                inner: CallPath {
+                    prefixes: vec![],
+                    suffix: Ident::new_no_span("decode_first_param".into()),
+                    is_absolute: false,
+                },
+                type_arguments: TypeArgs::Regular(vec![TypeArgument {
+                    type_id: string_slice_type_id,
+                    initial_type_id: string_slice_type_id,
+                    span: Span::dummy(),
+                    call_path_tree: None,
+                }]),
+                span: Span::dummy(),
+            },
+            arguments: vec![],
+        })),
+        span: Span::dummy(),
+    }
+}
+
+fn call_decode_second_param(_engines: &Engines, args_type: TypeArgument) -> Expression {
+    Expression {
+        kind: ExpressionKind::FunctionApplication(Box::new(FunctionApplicationExpression {
+            call_path_binding: TypeBinding {
+                inner: CallPath {
+                    prefixes: vec![],
+                    suffix: Ident::new_no_span("decode_second_param".into()),
+                    is_absolute: false,
+                },
+                type_arguments: TypeArgs::Regular(vec![args_type]),
+                span: Span::dummy(),
+            },
+            arguments: vec![],
+        })),
+        span: Span::dummy(),
+    }
+}
+
+fn call_eq(_engines: &Engines, l: Expression, r: Expression) -> Expression {
+    Expression {
+        kind: ExpressionKind::MethodApplication(Box::new(MethodApplicationExpression {
+            method_name_binding: TypeBinding {
+                inner: MethodName::FromModule {
+                    method_name: Ident::new_no_span("eq".to_string()),
+                },
+                type_arguments: TypeArgs::Regular(vec![]),
+                span: Span::dummy(),
+            },
+            contract_call_params: vec![],
+            arguments: vec![l, r],
+        })),
+        span: Span::dummy(),
+    }
+}
+
```

Comment:
This isn't directly relevant to the code I've marked, or indeed to the PR itself, but it did make me think: What's the semantics for passing parameters in contract calls? Is it always call-by-value (with copying), or can you do call-by-reference?

Reply:
Same question for values returned by contract calls, BTW: Can you return a reference, or does it have to be a value?

Reply:
We are "encoding" all arguments to bytes on the caller side, and decoding it on the "callee" side. So it is "call-by-value" that uses the new encoding to do a "deep copy", instead of a shallow `memcpy`.

Same thing for return values: "deep copy using the new encoding".

> Can you return a reference, or does it have to be a value?

Today the only type that does not have a default impl for AbiEncode/AbiDecode is `raw_pointer` because we dont want to pass pointers across contracts silently. 

Of course you can impl them yourself and decide to pass the pointer, but that means you would be inside "unsafe world".

Same thing for references. They will not be AbiEncode/AbiDecode by default.